### ISSUE TITLE
Significant lag when building mask

### ISSUE THREAD

**Issue by AipaMC** [IntellectualSites/PlotSquared#1307]
The mask building process here:
https://github.com/IntellectualSites/PlotSquared/blob/master/Core/src/main/java/com/plotsquared/listener/WEManager.java#L38
is very intensive because it has to iterate and process each plot in the world. This becomes very unreasonable when a world has over 100k plots.

WorldEditListener listens for the PlayerInteractEvent, which then creates EditSessions. The EditSessions call the listener in WELIstener and build the mask.
https://gyazo.com/6e045248574297fb87bf2154a99aef3e

This may be only an issue with AsyncWorldEdit, as it seems to create new EditSessions for each action (sometimes happening a few times per tick when a lot of people are on). 

Here is a stacktrace from when my server crashed from this:
http://pastebin.com/VJX4vNxV
and another: 
http://pastebin.com/j7aGQASB

Perhaps some sort of mask caching system? Maybe only rebuild the mask when the user changes world? Or perhaps just get the plots nearest to the player and process those instead of processing the plots from the entire world?

If you need any additional information, feel free to ask!

Edit:
Running the lastest version as of now (commit: daea9cf60de3aab92d3c74f934496760c0b5006c)

**Comment by AipaMC** [comment]
I was able to reduce the impact of the problem slightly by moving this permission check out of the loop since it is redundant.
https://github.com/IntellectualSites/PlotSquared/blob/master/Core/src/main/java/com/plotsquared/listener/WEManager.java#L55

This reduces the problem slightly, but doesn't come close to solving it. After some more testing I can confirm that the problem is isolated to using PS with AsyncWorldEdit. I've contacted the dev of AWE to see about reducing the amount of EditSessions that are created, however that does not change the fact that scanning the entire world's plots is a very intensive and unreasonable thing to do.
